### ISSUE TITLE
[bug] []string is not decoded correctly

### ISSUE THREAD

**Issue by napei** [gorilla/schema#152]
**Describe the bug**
> A clear and concise description of what the bug is.

Decoding a parameter into a string slice does not split on commas as it's supposed to, instead it creates a slice of one string, which is the input.

**Versions**
Go version: `go version go1.15 windows/amd64`
package version: `1.2.0` & versions before

**Steps to Reproduce**

See code below

**Expected behavior**

The value `test` should decode into an array of length 4.

**Code Snippets**

```go
package main

import (
	"encoding/json"
	"fmt"

	"github.com/gorilla/schema"
)

func main() {

	src := map[string][]string{
		"test": {"this,is,a,test"},
	}

	var out struct {
		Test []string `schema:"test"`
	}

	if err := schema.NewDecoder().Decode(&out, src); err != nil {
		panic(err)
	}

	j, _ := json.Marshal(out)
	fmt.Println(string(j))
	fmt.Println(len(out.Test))
}
```
This yields 
```
{"Test":["this,is,a,test"]}
1
```

I have marshalled to JSON as go omits quotes from it's prints which masks the problem

`[]bool`, `[]int` and other non-strings work fine

A small bit of debugging shows it's getting into this condition https://github.com/gorilla/schema/blob/master/decoder.go#L280 and not the last else block which it's supposed to for this particular case.

**Comment by Fenny** [comment]
I find all the reflection rather complex to make a PR/fix myself with the limited time I have. 
By any chance, you could take a quick look @elithrar ?

**Comment by s7v7nislands** [comment]
> Decoding a parameter into a string slice does not split on commas as it's supposed to.
why should split on commas?

**Comment by napei** [comment]
> why should split on commas?

`param=this,should,be,an,array` -> `[]string{"this","should","be","an","array"}`

**Comment by s7v7nislands** [comment]
```
package main

import (
	"fmt"
	"net/url"
)

func main() {

	q := "param=this,should,be,an,array"

	v, err := url.ParseQuery(q)
	if err != nil {
		fmt.Printf("error: %v", err)
		return
	}

	fmt.Printf("%+v\n", v)
	fmt.Printf("%d\n", len(v["q"]))



➜  tt go run main7.go
map[param:[this,should,be,an,array]]
1

```
I think this is not a schema bug. query string would't split on commas.

**Comment by napei** [comment]
> I think this is not a schema bug. query string would't split on commas.

`url.ParseQuery` uses different logic. The docs for gorilla/schema specify decoding of `a slice or a pointer to a slice of one of the above types` of which one is a string. Providing params as `param=this&param=should&param=be&param=an&param=array` works but every other type splits on commas. 

If strings not splitting on commas is intended, specifically decoding a param to `[]string` then it should be documented. Otherwise, it should be changed.

**Comment by stinkyfingers** [comment]
This bug seems like a feature. The tests specifically assert that the schema parser does not split strings on commas here: https://github.com/gorilla/schema/blob/master/decoder_test.go#L1165. The tests also provide an example of using `RegisterConverter` to provide a user-defined conversion function to the decoder here: https://github.com/gorilla/schema/blob/master/decoder_test.go#L1252. 

I imagine the difficult decision regarding whether to split on commas for `[]string` arises from the fact that commas _are_ strings (i.e. are they a delimiter or part of the data?) There's probably room for this to be documented more clearly.

TLDR;
Add a custom converter to avoid the "bug" in the above code snippet: 
```
package main

import (
	"encoding/json"
	"fmt"

	"github.com/gorilla/schema"
)

func main() {

	src := map[string][]string{
		"test": {"this,is,a,test"},
	}

	var out struct {
		Test []string `schema:"test"`
	}

	decoder := schema.NewDecoder()
	decoder.RegisterConverter([]string{}, func(input string) reflect.Value {
		return reflect.ValueOf(strings.Split(input, ","))
	})
	if err := decoder.Decode(&out, src); err != nil {
		panic(err)
	}

	j, _ := json.Marshal(out)
	fmt.Println(string(j))
	fmt.Println(len(out.Test))
}
```
